core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala

```diff
@@ -614,6 +621,22 @@ private[worker] class Worker(
     webUi.stop()
     metricsSystem.stop()
   }
+
+  private def trimFinishedExecutorsIfNecessary(): Unit = {
+    if (finishedExecutors.size > retainedExecutors) {
+      finishedExecutors.take(math.max(finishedExecutors.size / 10, 1)).foreach{
```

Comment:
I was about to work on this same issue until finding this PR already posted. I had one observation that I wanted to bring up for discussion. `scala.collection.concurrent.HashMap` does not preserve insertion ordering when you perform operations on it for iterator or traversal methods. So `take`, for example may be able to remove recent additions where the user might prefer to just lose the oldest executors from the list. [LinkedHashMap](http://www.scala-lang.org/api/2.11.5/index.html#scala.collection.mutable.LinkedHashMap) does provide the guarantee on insertion ordering being preserved for its operations. It comes at the cost of more memory overhead to provide this guarantee, but it may be worth it.